shim/src/main.rs

```diff
@@ -0,0 +1,144 @@
+mod package_manager;
+mod paths;
+
+use crate::package_manager::PackageManager;
+use crate::paths::AncestorSearch;
+use anyhow::Result;
+use clap::{Parser, Subcommand};
+use std::path::Path;
+use std::{
+    env,
+    ffi::CString,
+    os::raw::{c_char, c_int},
+    process,
+};
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None, ignore_errors = true, disable_help_flag = true)]
+struct Args {
+    /// Current working directory
+    #[clap(long, value_parser)]
+    cwd: Option<String>,
+    #[clap(subcommand)]
+    commands: Option<Commands>,
+    task: Option<String>,
+}
+
+/// Defines the subcommands for CLI. NOTE: If we change the commands in Go,
+/// we must change these as well to avoid accidentally passing the --single-package
+/// flag into non-build commands.
+#[derive(Subcommand, Debug)]
+enum Commands {
+    Bin,
+    Completion,
+    Daemon,
+    Help,
+    Link,
+    Login,
+    Logout,
+    Prune,
+    Unlink,
+    Run { tasks: Vec<String> },
+}
+
+extern "C" {
+    pub fn nativeRunWithArgs(argc: c_int, argv: *mut *mut c_char) -> c_int;
+}
+
+/// Runs the turbo in the current binary
+///
+/// # Arguments
+///
+/// * `args`: Arguments for turbo
+///
+/// returns: Result<i32, Error>
+///
+fn run_current_turbo(args: Vec<String>) -> Result<i32> {
+    let mut args = args
+        .into_iter()
+        .map(|s| {
+            let c_string = CString::new(s)?;
+            Ok(c_string.into_raw())
+        })
+        .collect::<Result<Vec<*mut c_char>>>()?;
+    args.shrink_to_fit();
+    let argc: c_int = args.len() as c_int;
+    let argv = args.as_mut_ptr();
+    let exit_code = unsafe { nativeRunWithArgs(argc, argv) };
+    Ok(exit_code)
+}
+
+/// Checks if we are in single package mode by first seeing if there is a turbo.json
```

Comment:
```suggestion
/// Checks if we are in "single package mode" by first seeing if there is a turbo.json
```

When we have to scatter a feature's implementation around the codebase, it's helpful to give it a name or some kind of searchable string so the next time someone needs to work on anything related to single package mode, they can find all the references to it more easily. My diff suggestion doesn't make that more likely, but it does make it more clear that it's a named feature. If you have a better way to solve this problem, go for it!